docs/framework-integrations/react.mdx

```diff
@@ -42,62 +42,131 @@ You also need to install the UI plugin you want to use. For instance,
 other UI elements. The components can be used with either [React][] or
 API-compatible alternatives such as [Preact][].
 
-The following plugins are available as React component wrappers:
+:::caution
+
+If you find yourself writing many instances of `useState` and `useEffect` to
+achieve something with Uppy in React, you are most likely breaking React best
+practices. Consider reading
+“[You Might Not Need an Effect](https://react.dev/learn/you-might-not-need-an-effect)”
+and looking at our examples below.
+
+:::
+
+### Components
+
+The following components are exported from `@uppy/react`:
 
 - `<Dashboard />` renders [`@uppy/dashboard`](/docs/dashboard)
 - `<DragDrop />` renders [`@uppy/drag-drop`](/docs/drag-drop)
 - `<ProgressBar />` renders [`@uppy/progress-bar`](/docs/progress-bar)
 - `<StatusBar />` renders [`@uppy/status-bar`](/docs/status-bar)
 
-:::caution
+### Hooks
 
-A couple things to keep in mind when using Uppy with React:
+`useUppyState(uppy, selector)`
 
-- Instead of adding a UI plugin to an Uppy instance with `.use()`, the Uppy
-  instance can be passed into components as an `uppy` prop.
-- All other props are passed as options to the plugin.
-- The Uppy instance **should not live inside the component** but outside of it
-  (for class components, it should not be instantiated inside the `render()`
-  method).
-- You have to pass the IDs of your `use`d plugins to the `plugins` array prop so
-  Dashboard knows it needs to render them.
-- An Uppy instance can’t be used by more than one component. Make sure you are
-  using a unique Uppy instance per component.
+Use this hook when you need to access Uppy’s state reactively. Most of the
+times, this is needed if you are building a custom UI for Uppy in React.
 
-:::
+```js
+// IMPORTANT: passing an initializer function to prevent the state from recreating.
+const [uppy] = useState(() => new Uppy());
+
+const files = useUppyState(uppy, (state) => state.files);
+const totalProgress = useUppyState(uppy, (state) => state.totalProgress);
+// We can also get specific plugin state.
+// Note that the value on `plugins` depends on the `id` of the plugin.
+const metaFields = useUppyState(
+	uppy,
+	(state) => state.plugins?.Dashboard?.metaFields,
+);
+```
```

Comment:
This can be abstracted one level deeper to offer a better DX, while avoiding prop drilling with the `uppy` variable:

```jsx
const UppyContext = createContext(null)

function useUppyState(selector) {
  const uppy = useContext(UppyContext)
  if (!uppy) throw new Error('needs to be in context')
  return whateverUseUppyStateDidBefore(uppy, selector)
}

function useUppy() {
  return useContext(UppyContext)
}

function UppyProvider(props) {
  const uppy = useRef()

  if (!uppy.current) {
    uppy.current = new Uppy(props.options)
  }

  return (
    <UppyContext.Provider value={uppy.current}>
      {props.children}
    </UppyContext.Provider>
  )
}
```

This would allow folks:

- to wrap each component that needs its own instance of Uppy with the provider
- to call `useUppyState` anywhere in the tree without having to pass `uppy` down
- to update Uppy's options in `useEffect` hooks or event handlers from inside the tree by using `useUppy`

Reply:
I'll copy my comment from the PR itself as I still stand by it.

---

I thought about it but refrained from doing so:

> Context is very tempting to use! However, this also means it’s too easy to overuse it. **Just because you need to pass some props several levels deep doesn’t mean you should put that information into context** — [React docs](https://react.dev/learn/passing-data-deeply-with-context#before-you-use-context)

The React docs use cases for adding context also don't align with what Uppy is used for. Realistically, there is only one page where people upload in most apps and it would be overkill to provide a context for this in my opinion. I'd rather keep it simple and explicit for now. Perhaps reconsidering when we see people needing this more.